File: leetcode/421.数组中两个数的最大异或值.go
func findMaximumXOR(nums []int) int {
    ans := 0
    highBit := bits.Len(uint(slices.Max(nums))) - 1
    vis := map[int]bool{}
    mask := 0
    for i := highBit; i >= 0; i-- {
        clear(vis)
        mask |= 1 << i
        newAns := ans | 1 << i
        for _, x := range nums {
            x &= mask
            if vis[newAns ^ x] {
                ans = newAns
                break
            }
            vis[x] = true
        }
    }
    return ans
}

File: leetcode/421.数组中两个数的最大异或值.py
class Solution:
    def findMaximumXOR(self, nums: List[int]) -> int:
        ans = mask = 0
        # 拆位计算
        highbit = max(nums).bit_length() - 1
        for i in range(highbit, -1, -1):
            mask |= 1 << i
            new_ans = ans | (1 << i)
            vis = set()
            for x in nums:
                new_x = x & mask
                if new_x ^ new_ans in vis:
                    ans = new_ans
                    break
                vis.add(new_x)
        return ans
